File: Neato-SDK/neato-sdk-android/src/test/java/com/neatorobotics/sdk/android/authentication/DefaultAccessTokenDatasourceTest.java
package com.neatorobotics.sdk.android.authentication;

import android.content.Context;
import android.content.SharedPreferences;
import android.net.Uri;
import android.test.suitebuilder.annotation.SmallTest;

import org.junit.Before;
import org.junit.Test;
import org.junit.runner.RunWith;
import org.mockito.Mock;
import org.mockito.MockitoAnnotations;
import org.mockito.runners.MockitoJUnitRunner;

import java.util.Calendar;
import java.util.Date;
import java.util.Map;

import static org.junit.Assert.assertEquals;
import static org.junit.Assert.assertNotNull;
import static org.junit.Assert.assertNull;
import static org.junit.Assert.assertTrue;
import static org.mockito.Matchers.anyInt;
import static org.mockito.Matchers.anyString;
import static org.mockito.Mockito.when;

@RunWith(MockitoJUnitRunner.class)
@SmallTest
public class DefaultAccessTokenDatasourceTest {

    @Mock
    SharedPreferences sharedPreferences;

    @Mock
    SharedPreferences.Editor editor;

    @Mock
    Context context;

    @Before
    public void setup() {
        MockitoAnnotations.initMocks(this);
        when(sharedPreferences.getString(anyString(), anyString())).thenReturn("value");
        when(sharedPreferences.edit()).thenReturn(editor);
        when(context.getSharedPreferences(anyString(), anyInt())).thenReturn(sharedPreferences);
    }

    @Test
    public void formatHTTP11DateString() throws Exception {
        Date now = new Date();
        String formattedDate = DefaultAccessTokenDatasource.serializeDate(now);
        assertNotNull(formattedDate);
    }

    @Test
    public void parseHTTP11DateString() throws Exception {
        String inputStr = "2014-01-02T12:00:00Z";
        Date parsedDate= DefaultAccessTokenDatasource.deserializeDate(inputStr);
        assertNotNull(parsedDate);
    }

    @Test
    public void parseAndFormatHTTP11DateString() throws Exception {
        String inputStr = "2014-01-02T12:00:00Z";
        String inputStr2 = "2014-01-02T12:00:01Z";
        Date parsedDate= DefaultAccessTokenDatasource.deserializeDate(inputStr);
        assertNotNull(parsedDate);

        String formattedStr = DefaultAccessTokenDatasource.serializeDate(parsedDate);
        assertEquals(inputStr, formattedStr);

        Date parsedDate2= DefaultAccessTokenDatasource.deserializeDate(inputStr2);

        Calendar cal1 = Calendar.getInstance();
        cal1.setTime(parsedDate);

        Calendar cal2 = Calendar.getInstance();
        cal2.setTime(parsedDate2);
        assertTrue(cal1.before(cal2));
    }
}

File: Neato-SDK/neato-sdk-android/src/test/java/com/neatorobotics/sdk/android/authentication/DefaultAccessTokenDatasourceTest.kt
package com.neatorobotics.sdk.android.authentication

import android.content.Context
import android.content.SharedPreferences

import org.junit.Before
import org.junit.Test
import org.junit.runner.RunWith
import org.mockito.Mock
import org.mockito.MockitoAnnotations
import org.mockito.junit.MockitoJUnitRunner

import java.util.Calendar
import java.util.Date

import org.junit.Assert.assertEquals
import org.junit.Assert.assertNotNull
import org.junit.Assert.assertTrue

@RunWith(MockitoJUnitRunner::class)
class DefaultAccessTokenDatasourceTest {

    @Mock
    internal var sharedPreferences: SharedPreferences? = null

    @Mock
    internal var editor: SharedPreferences.Editor? = null

    @Mock
    internal var context: Context? = null

    @Before
    fun setup() {
        MockitoAnnotations.initMocks(this)
    }

    @Test
    fun formatHTTP11DateString() {
        val now = Date()
        val formattedDate = DefaultAccessTokenDatasource.serializeDate(now)
        assertNotNull(formattedDate)
    }

    @Test
    fun parseHTTP11DateString() {
        val inputStr = "2014-01-02T12:00:00Z"
        val parsedDate = DefaultAccessTokenDatasource.deserializeDate(inputStr)
        assertNotNull(parsedDate)
    }

    @Test
    fun parseAndFormatHTTP11DateString() {
        val inputStr = "2014-01-02T12:00:00Z"
        val inputStr2 = "2014-01-02T12:00:01Z"
        val parsedDate = DefaultAccessTokenDatasource.deserializeDate(inputStr)
        assertNotNull(parsedDate)

        val formattedStr = DefaultAccessTokenDatasource.serializeDate(parsedDate)
        assertEquals(inputStr, formattedStr)

        val parsedDate2 = DefaultAccessTokenDatasource.deserializeDate(inputStr2)

        val cal1 = Calendar.getInstance()
        cal1.time = parsedDate

        val cal2 = Calendar.getInstance()
        cal2.time = parsedDate2
        assertTrue(cal1.before(cal2))
    }
}